### PR TITLE
Memory management for array allocation in nopython

### PR DESCRIPTION
**Pull request by sklam**
* Introduces a *numba runtime* or NRT
* New option to jit, njit ``nrt=True`` to enable NRT.
* support ``np.empty()`` in nopython+nrt
* return array, return slices of array in nopython+nrt
* return any array in nopython+nrt

### CODE REVIEW

**Comment by ufechner7** [comment]
How does this work? Do you use the memory allocator of Python, or a separate one? Do you use reference counting or garbage collection?

**Comment by sklam** [comment]
This implementation uses a custom atomic reference counting.  The atomic operations are generated by LLVM for portability so you won't see it in the C code.  Cyclic reference is not yet handled.  Our first and major usecase is for arrays, which cannot have cyclic reference.  A separate reference count is maintained by NRT so that incref/decref do not need the GIL.  

At https://github.com/sklam/numba/commit/f75cdfd548474caf994da5dc15a5ef3d9c3fc287, the implementation is not completely threadsafe yet.  Internally, NRT maintains some data structures (the freelist for ``MemInfo``) that are subject race condition.  I am working on a threadsafe version by using atomic compare-and-swap.

File: numba/numpy_support.py [hunk 1]
```diff
@@ -123,6 +123,8 @@ def as_dtype(nbtype):
     if isinstance(nbtype, (types.CharSeq, types.UnicodeCharSeq)):
         letter = _as_dtype_letters[type(nbtype)]
         return numpy.dtype('%s%d' % (letter, nbtype.count))
+    if isinstance(nbtype, types.Record):
```
Comment (gmarkall) [inline numba/numpy_support.py]: I don't think this is required - I think it would only be hit if you had a record within a record, which is not supported at the moment.

File: numba/tests/test_dyn_array.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+        cfunc = nrtjit(return_external_array)
+        out = cfunc()
+
+        yy = y[2:]
+        np.testing.assert_equal(yy, out)
+        np.testing.assert_equal(yy, np.ones(2, dtype=np.float32))
+        np.testing.assert_equal(out, np.ones(2, dtype=np.float32))
+
+    def test_array_pass_through(self):
```
Comment (gmarkall) [inline numba/tests/test_dyn_array.py]: Would it be meaningful to test that the refcount of the array is unchanged by calling the jitted function here?

File: numba/lowering.py [hunk 3]
```diff
@@ -247,7 +257,13 @@ def lower_inst(self, inst):
             return impl(self.builder, (target, index, value))
 
         elif isinstance(inst, ir.Del):
-            pass
+            try:
+                # XXX: incorrect Del injection?
+                val = self.loadvar(inst.value)
```
Comment (gmarkall) [inline numba/lowering.py]: IIUC, loadvar is creating a new load instruction that is not used - if this is the case, would calling getvar be enough to determine whether the `KeyError` is thrown?

File: numba/numpy_support.py (hunk 1 above)
Comment (sklam) [inline numba/numpy_support.py]: This is not a required change.  Just that in some of my exploration, I wanted to call `as_dtype(my_numba_record)` for creating a a structure dtype.  Also, this will make `my_numba_record==from_dtype(as_dtype(my_numba_record))` work.

File: numba/tests/test_dyn_array.py (hunk 2 above)
Comment (sklam) [inline numba/tests/test_dyn_array.py]: Yes, it will be good to check our refcount behavior.  Here, `y` is referenced by the numba IR which is kept alive by the dispatcher.

File: numba/lowering.py (hunk 3 above)
Comment (sklam) [inline numba/lowering.py]: Yes, `getvar` will be enough.  The extra load wouldn't be a problem.  LLVM will optimize it out.  Ultimately, I want to improve the `Del` injection so I don't need the try-except here.

File: numba/tests/test_dyn_array.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+        def return_external_array():
+            return y
+
+        cfunc = nrtjit(return_external_array)
+        out = cfunc()
+
+        # out reference by cfunc
+        self.assertEqual(initrefct + 1, sys.getrefcount(y))
+
+        np.testing.assert_equal(y, out)
```
Comment (pitrou) [inline numba/tests/test_dyn_array.py]: Is `out` the same object as `y`? Is it backed by the same memory? If yes, then test mutating one effects the other as well?

File: numba/_arraystruct.h [hunk 5]
```diff
@@ -0,0 +1,21 @@
+#ifndef NUMBA_ARYSTRUCT_H_
+#define NUMBA_ARYSTRUCT_H_
+/*
+ * Fill in the *arystruct* with information from the Numpy array *obj*.
+ * *arystruct*'s layout is defined in numba.targets.arrayobj (look
+ * for the ArrayTemplate class).
+ */
+
+typedef struct {
+    void     *meminfo;
```
Comment (pitrou) [inline numba/_arraystruct.h]: Add a comment explaining where "meminfo" is described perhaps?

File: numba/tests/test_dyn_array.py [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+import numpy as np
+import threading
+
+from numba import unittest_support as unittest
+from numba import njit
+from numba import utils
+
+
+
+nrtjit = njit(nrt=True)
```
Comment (pitrou) [inline numba/tests/test_dyn_array.py]: I don't think it makes sense to expose "nrt" to the user. Numba should always allow memory allocation from CPU funcs, IMO.

File: numba/runtime/_nrt_python.c [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+
+    if (miobj) {
+        /* Set the MemInfoObject as the base object */
+        PyArray_SetBaseObject((PyArrayObject*)array, (PyObject *)miobj);
+    }
+    return array;
+}
+
+static void
+NRT_adapt_buffer(Py_buffer *buf, arystruct_t *arystruct)
```
Comment (pitrou) [inline numba/runtime/_nrt_python.c]: Does it have to be separate from the equivalent function in _helperlib.c?

**Comment by pitrou** [comment]
So, from a quick glance, the implementation adds a lot of complication that doesn't sound strictly necessary for a reference-counting scheme:
- the reference count is a separately allocated structure rather than an inline member
- it has its own freelist scheme
- it has a "deferred destructor" scheme (what is that, and why is it useful?)
- it has its own spinlock implementation (this sounds like a bad idea, IMHO)
- it maintains function pointers for the atomic increment and decrement functions

Ideally:
- the reference count field is all we have to add to existing structures, and it can be inlined (similar to PyObject_HEAD for CPython objects)
- the object type is known at compile time, so is the destructor
- there is no need to customize the increment and decrement functions at runtime
- if we want to handle lock-free structures, it sounds better to use an existing library :-)

File: numba/tests/test_dyn_array.py (hunk 4 above)
Comment (sklam) [inline numba/tests/test_dyn_array.py]: `out` is not `y` in this code.  I haven't change the semantic of global array, yet.  I think it should be a separate pull request.

File: numba/tests/test_dyn_array.py (hunk 6 above)
Comment (sklam) [inline numba/tests/test_dyn_array.py]: I think having `nrt=True` to be the default and allowing it to be disabled could be helpful for numba developers for testing and debugging.  We can make it `_nrt` to suggesting it is internal usage.  What do you think?

File: numba/runtime/_nrt_python.c (hunk 7 above)
Comment (sklam) [inline numba/runtime/_nrt_python.c]: No, this is temporary.

Ultimately, I want 'nrt' be a shared library that has no python dependency for potential pycc use cases.  So that, pycc can compile standalone binaries that are linked nrt.  I see _nrt_python to be the bridge that provides the adaptors between nrt and python.

Thoughts?

Comment (pitrou) [inline numba/runtime/_nrt_python.c]: Is pycc useful? We don't document it and we get no reports of its use in the wild.

Comment (seibert) [inline numba/runtime/_nrt_python.c]: pycc is not currently useful (it's functionally was fundamentally broken in the Great Numba Refactor), but we do get requests from people for the ability to compile code such that the Numba module is no longer needed.

We have several things to resolve before that is possible, but I think we do want to plan for how nopython mode code can be made independent of Numba.

**Comment by sklam** [comment]
> So, from a quick glance, the implementation adds a lot of complication that doesn't sound strictly necessary for a reference-counting scheme:

> the reference count is a separately allocated structure rather than an inline member

In the emitted llvm, numba treats most data as immutable.  Structures are passed around within each function as LLVM structs.  To have a mutable field, we will need to make structures pass by reference (by pointer to the LLVM struct).  My implementation makes the MemInfo an opaque pointer that can be treated as an immutable field.  This avoids substantial changes to the compiler.

As to why we pass structures by value:
* it is faster (due to fewer indirection) on GPU architectures where memory bandwidth are scarce.
* easier ABI for functions since we only pass around non-aggregate, primitives types (and there are so few of them).

> it has its own freelist scheme

This is to avoid too much malloc/free calls.  But, I haven't proven it to be useful.  

> it has a "deferred destructor" scheme (what is that, and why is it useful?)

Memory backed by a PyObject requires the GIL for Py_DecRef, which is called once the NRT refct drops to zero.  I am considering if it would be useful to allow the destructor be deferred so that we can call perform multiple Py_DecRef in a GIL acquire+release.  Otherwise, we are calling one Py_DecRef per GIL acquire+release.

> it has its own spinlock implementation (this sounds like a bad idea, IMHO)

I am using LLVM to generate the atomic operations for portability reason.  That way I don't have to change anything when we port to a new architecture.  Of course, we can always substitute it with a portable atomic+lock library.

> it maintains function pointers for the atomic increment and decrement functions
see comment above.

> Ideally:

> the reference count field is all we have to add to existing structures, and it can be inlined (similar to PyObject_HEAD for CPython objects)

see comment above

> the object type is known at compile time, so is the destructor

We only differentiate between NRT allocated vs Python allocated at this time.  The compiler needs to insert type-specific destructor for types containing other allocated objects.  But, a function does not know who allocated the memory.

> there is no need to customize the increment and decrement functions at runtime
> if we want to handle lock-free structures, it sounds better to use an existing library :-)

Can you suggest a library?

**Comment by pitrou** [comment]
> In the emitted llvm, numba treats most data as immutable. Structures are passed around within each function as LLVM structs

Oh, yes, sorry for that. I always forget it (my mistake for often dealing with PyObject pointers :-)).

That said, the MemInfo could still be stack-allocated in the Numba function (a bit like generators), rather than heap-allocated. I wonder if that would be complicated to implement...

> Memory backed by a PyObject requires the GIL for Py_DecRef, which is called once the NRT refct drops to zero.

I don't understand how a Numba-compiled function can ever call Py_DecRef in nopython mode. That means the object was passed as parameter to the function, but then we shouldn't need to memory-manage it ourselves...

> I am using LLVM to generate the atomic operations for portability reason

I am talking about the lock / unlock primitives in the NRT lib. AFAIU, tuning a spinlock is delicate (you want to yield after a certain # of failures, otherwise you may waste a whole timeslice waiting for the unlock).

By the way, if the freelist is just a single-linked list, perhaps you can just atomically replace the list head without any locking? Something like:

```
list *newhead = malloc(...);
do {
    list *head = global_head;   /* volatile */
    newhead->next = head;
} while (!atomic_cas(&global_head, head, newhead));
```

> Can you suggest a library?

I'll search for one if this is really necessary. But I wonder if we could simply not have any spinlock at all.

**Comment by sklam** [comment]
> I don't understand how a Numba-compiled function can ever call Py_DecRef in nopython mode. That means the object was passed as parameter to the function, but then we shouldn't need to memory-manage it ourselves...

It can happen with container types once we have them. 

> By the way, if the freelist is just a single-linked list, perhaps you can just atomically replace the list head without any locking? Something like:

Yes, I will try that.

**Comment by pitrou** [comment]
> It can happen with container types once we have them. 

IMO that's a bit prematurate :-) Also, if we want a specialized representation of container types, not a bunch of PyObject pointers, we wouldn't share the representation with the CPython one...

**Comment by sklam** [comment]
It is for the case when nrt is borrowing memory owned by a pyobject.

Let's say we have:

```
class Foo(object):
    def __init__(self, array):
        self.array = array
```

If we don't want to copy the data of array (a numpy.ndarray), MemInfo needs to keep a reference to the original object.  Decref on MemInfo may trigger a Py_Decref on the original pyobject.

**Comment by sklam** [comment]
> IMO that's a bit prematurate :-) 

I agree.  It is not fully developed.  It wasn't turn on by default.  I just wanted to make sure it is possible.  
It probably need to trigger the cleanup once awhile to avoid accumulating too many deferred dtors.

**Comment by seibert** [comment]
Note that the immediate followup to this work is going to be working on JIT classes, so I don't think support for retention of arrays inside a container is that premature.  We certainly want to think through that case before we commit to this approach to memory management.

File: numba/runtime/nrt.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        _nrt.memsys_set_atomic_cas(self._ptr_cas)
+
+    def __del__(self):
+        _nrt.memsys_shutdown()
+
+    def meminfo_new(self, data, pyobj):
+        mi = _nrt.meminfo_new(data, pyobj)
+        return MemInfo(mi)
+
+    def meminfo_alloc(self, size, safe=False):
```
Comment (gmarkall) [inline numba/runtime/nrt.py]: `safe` here appears to mean "allocated memory is zero-filled" (if I've understood correctly) - perhaps this should be documented or renamed to something more descriptive? (or is "safe" == "zero-filled" a convention?)

Comment (sklam) [inline numba/runtime/nrt.py]: I added this during debugging.  I wanted it to mean anything that makes allocation safer.  All I could think of is zero-filling after malloc() and before free().  It helped track down premature deallocation.  I will document what it does.  We can enhance it to do more.  I don't expect it to be enabled for optimized run.

File: numba/lowering.py [hunk 9]
```diff
@@ -185,8 +185,18 @@ def lower_inst(self, inst):
             self.context.debug_print(self.builder, str(inst))
         if isinstance(inst, ir.Assign):
             ty = self.typeof(inst.target.name)
+            try:
```
Comment (pitrou) [inline numba/lowering.py]: Should this additional code be put in storevar() instead?

File: numba/pythonapi.py [hunk 10]
```diff
@@ -1201,6 +1217,31 @@ def from_native_generator(self, val, typ, env=None):
...
+
+        ndim = self.context.get_constant(types.int32, aryty.ndim)
+        typenum = self.context.get_constant(types.int32, dtype.num)
+
+        aryptr = cgutils.alloca_once_value(self.builder, ary)
+        return self.builder.call(fn, [self.builder.bitcast(aryptr,
+                                                           self.voidptr),
+                                      ndim, typenum])
+
+    def nrt_array_adaptor(self, ary, ptr):
```
Comment (pitrou) [inline numba/pythonapi.py]: The difference between "nrt_array_adaptor" and "nrt_adapt_native_array" is not obvious at first sight :-)

File: numba/runtime/_nrt_python.c [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+        return NULL;
+    }
+    NRT_MemSys_shutdown();
+    Py_RETURN_NONE;
+}
+
+static
+PyObject*
+memsys_set_atomic_inc_dec(PyObject *self, PyObject *args) {
+    unsigned PY_LONG_LONG addr_inc, addr_dec;
```
Comment (pitrou) [inline numba/runtime/_nrt_python.c]: I'm not sure PY_LONG_LONG is always enough for a pointer, but it probably is on common architectures... Otherwise, the right way would be PyLong_AsVoidPtr().

File: numba/runtime/_nrt_python.c [hunk 12]
```diff
@@ -0,0 +1,539 @@
...
+    Py_RETURN_NONE;
+}
+
+
+static
+void pyobject_dtor(void *ptr, void* info) {
+    PyGILState_STATE gstate;
+    PyObject *ownerobj = info;
+
+    gstate = PyGILState_Ensure();   /* ensure the GIL */
```
Comment (pitrou) [inline numba/runtime/_nrt_python.c]: If the destructors are always "deferred", you shouldn't need this (i.e. the destructor is called either from the wrapper or Python itself).

File: numba/runtime/_nrt_python.c [hunk 13]
```diff
@@ -0,0 +1,539 @@
...
+
+/*
+ * Create a new MemInfo with a new NRT allocation
+ */
+static
+PyObject*
+meminfo_alloc(PyObject *self, PyObject *args) {
+    MemInfo *mi;
+    Py_ssize_t size;
+    if (!PyArg_ParseTuple(args, "K", &size)) {
```
Comment (pitrou) [inline numba/runtime/_nrt_python.c]: The code for Py_ssize_t is "n", not "K".

File: numba/runtime/_nrt_python.c [hunk 14]
```diff
@@ -0,0 +1,539 @@
...
+    MemInfoObject *mio = (MemInfoObject *)self;
+    MemInfo *mi = mio->meminfo;
+    if (lenp) {
+        *lenp = NRT_MemInfo_size(mi);
+    }
+    return 1;
+}
+
+#if (PY_MAJOR_VERSION < 3)
+static PyBufferProcs MemInfo_bufferProcs = {MemInfo_rdwrbufferproc,
```
Comment (pitrou) [inline numba/runtime/_nrt_python.c]: Instead of that, you can implement the new buffer protocol on Python 2. Enable the Py_TPFLAGS_HAVE_NEWBUFFER and see here for an example:
https://hg.python.org/cpython/file/4bf210f59ac6/Objects/bytearrayobject.c#l2800

File: numba/runtime/_nrt_python.c [hunk 15]
```diff
@@ -0,0 +1,539 @@
...
+static
+PyObject*
+MemInfo_release(MemInfoObject *self) {
+    NRT_MemInfo_release(self->meminfo, self->defer);
+    Py_RETURN_NONE;
+}
+
+static
+PyObject*
+MemInfo_release_defer(MemInfoObject *self) {
```
Comment (pitrou) [inline numba/runtime/_nrt_python.c]: Do we really need a separate method here? This is just for testing, right?

File: numba/runtime/_nrt_python.c [hunk 16]
```diff
@@ -0,0 +1,539 @@
...
+     "Increment the reference count"
+    },
+    {"release", (PyCFunction)MemInfo_release, METH_NOARGS,
+     "Decrement the reference count"
+    },
+    {"release_defer", (PyCFunction)MemInfo_release_defer, METH_NOARGS,
+     "Decrement the reference count but defer the destructor regardless"
+     "of the value in defer attribute"
+    },
+    {"set_defer", (PyCFunction)MemInfo_set_defer, METH_VARARGS,
```
Comment (pitrou) [inline numba/runtime/_nrt_python.c]: Just for the record, set_defer / get_defer could have been a "getset" instead (there's an example in numba/npyufunc/_internal.c).

File: numba/runtime/atomicops.py [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+    _define_atomic_inc_dec(module, "sub", ordering='monotonic')
+    _define_atomic_cmpxchg(module, ordering='monotonic')
+    return module
+
+
+def compile_atomic_ops():
+    # Implement LLVM module with atomic ops
+    mod = llvm.parse_assembly(str(define_atomic_ops()))
+
+    # Create a target just for this module
```
Comment (pitrou) [inline numba/runtime/atomicops.py]: Can't you reuse the codegen infrastructure here?

File: numba/runtime/nrt.c [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+
+/* The Memory System object */
+static MemSys TheMSys;
+
+
+static
+MemInfo *nrt_pop_meminfo_list(MemInfo * volatile *list) {
+    MemInfo *old, *repl, *head;
+
+    head = *list;     /* get the current head */
```
Comment (pitrou) [inline numba/runtime/nrt.c]: You certainly want this inside the loop (since another thread may change the head pointer).

File: numba/runtime/nrt.py [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    def __init__(self):
+        # Compile atomic operations
+        compiled = atomicops.compile_atomic_ops()
+        self._mcjit, self._ptr_inc, self._ptr_dec, self._ptr_cas = compiled
+        # Install atomic ops to NRT
+        _nrt.memsys_set_atomic_inc_dec(self._ptr_inc, self._ptr_dec)
+        _nrt.memsys_set_atomic_cas(self._ptr_cas)
+
+    def __del__(self):
+        _nrt.memsys_shutdown()
```
Comment (pitrou) [inline numba/runtime/nrt.py]: Since the "memsys" is a singleton, the Runtime object can only be used once, right? If so, I think it should be instantiated as a global in this module. Otherwise, it's too easy to make a mistake.

File: numba/runtime/nrt.c [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+
+void NRT_MemSys_set_atomic_cas_stub() {
+    NRT_MemSys_set_atomic_cas(nrt_testing_atomic_cas);
+}
+
+MemInfo* NRT_MemInfo_new(void *data, size_t size, dtor_function dtor,
+                         void *dtor_info)
+{
+    MemInfo * mi = NRT_MemSys_pop_meminfo();
+    mi->payload.refct = 0;
```
Comment (pitrou) [inline numba/runtime/nrt.c]: This feels a bit weird. You should start with a refcount of 1, no?

File: numba/runtime/atomicops.py [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+
+    _define_incref(module, incref)
+    _define_decref(module, decref)
+
+
+def _define_atomic_inc_dec(module, op, ordering):
+    """Define a llvm function for atomic increment/decrement to the given module
+    Argument ``op`` is the operation "add"/"sub".  Argument ``ordering`` is
+    the memory ordering.
+    """
```
Comment (pitrou) [inline numba/runtime/atomicops.py]: Please mention in the docstring that the function returns the new value (while LLVM's atomic_rmw returns the original value).

File: numba/runtime/atomicops.py [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+    ftype = ir.FunctionType(_word_type, [_word_type.as_pointer()])
+    fn_atomic = ir.Function(module, ftype, name="nrt_atomic_{0}".format(op))
+
+    [ptr] = fn_atomic.args
+    bb = fn_atomic.append_basic_block()
+    builder = ir.IRBuilder(bb)
+    ONE = ir.Constant(_word_type, 1)
+    builder.atomic_rmw(op, ptr, ONE, ordering=ordering)
+    res = builder.load(ptr)
```
Comment (pitrou) [inline numba/runtime/atomicops.py]: I think there is a race condition. If another thread runs in-between, then the two threads will see a resulting refcount of 0, so the destructor will be run twice. You want to reapply `op` to the result of the atomic_rmw instruction, instead (or you can change the decref test to check for 1, for example).

File: numba/runtime/atomicops.py [hunk 23]
```diff
@@ -0,0 +1,211 @@
...
+
+    return mcjit, atomic_inc, atomic_dec, atomic_cas
+
+
+_regex_incref = re.compile(r'call void @NRT_incref\((.*)\)')
+_regex_decref = re.compile(r'call void @NRT_decref\((.*)\)')
+_regex_bb = re.compile(r'[-a-zA-Z$._][-a-zA-Z$._0-9]*:')
+
+
+def remove_redundant_nrt_refct(ir_module):
```
Comment (pitrou) [inline numba/runtime/atomicops.py]: Actually, this is called with a LL module, not an IR module, right?

File: numba/targets/arrayobj.py [hunk 24]
```diff
@@ -1238,3 +1224,54 @@ def iternext_numpy_ndindex(context, builder, sig, args, result):
     nditer = nditercls(context, builder, value=nditer)
 
     nditer.iternext_specific(context, builder, result)
+
+
+@builtin
+@implement(numpy.empty, types.Kind(types.Integer))
+@implement(numpy.empty, types.Kind(types.Tuple))
```
Comment (pitrou) [inline numba/targets/arrayobj.py]: By the way, you can now use types.BaseTuple to match both Tuple and UniTuple.

File: numba/tests/test_dyn_array.py [hunk 25]
```diff
@@ -0,0 +1,362 @@
...
+            out = np.empty(inp.size)
+            tmp = 0
+            for i in range(out.size):
+                out[i] = tmp
+                tmp = inp[i]
+            return out
+
+        cfunc = nrtjit(pyfunc)
+        size = 10**5
+        arr = np.random.randint(0, 1, size)
```
Comment (pitrou) [inline numba/tests/test_dyn_array.py]: This creates a zero-filled array.

File: numba/tests/test_dyn_array.py (hunk 6 above)
Comment (pitrou) [inline numba/tests/test_dyn_array.py]: That would sound better, yes. Also, don't you want to add `nopython=True`?

File: numba/tests/test_dyn_array.py [hunk 26]
```diff
@@ -0,0 +1,362 @@
...
+            arr = np.random.randint(0, 1, size)
+            out = np.empty_like(arr)
+            thread = threading.Thread(target=wrapped,
+                                      args=(arr, out),
+                                      name="worker{0}".format(i))
+            workers.append(thread)
+            inputs.append(arr)
+            outputs.append(out)
+
+        # Launch worker threads
```
Comment (pitrou) [inline numba/tests/test_dyn_array.py]: I'm not sure this tests much. The threads will be serialized, since you didn't add `nogil=True`. Furthermore, they all operate on distinct arrays, so there's not much chance for contention (is there?).

File: numba/tests/test_dyn_array.py [hunk 27]
```diff
@@ -0,0 +1,362 @@
...
+        # Create worker threads
+        for i in range(10):
+            out = np.empty(size)
+            thread = threading.Thread(target=wrapped,
+                                      args=(size, repeat, out),
+                                      name="worker{0}".format(i))
+            workers.append(thread)
+            outputs.append(out)
+
+        # Launch worker threads
```
Comment (pitrou) [inline numba/tests/test_dyn_array.py]: Similarly, I'm not sure what you're testing here. As long as the threads are serialized and don't access the same objects, there's not much to test for :-) (the freelist perhaps?)

File: numba/tests/test_nrt.py [hunk 28]
```diff
@@ -0,0 +1,174 @@
...
+    alive = 0
+
+    def __init__(self):
+        type(self).alive += 1
+
+    def __del__(self):
+        type(self).alive -= 1
+
+
+class TestNrtMemInfo(unittest.TestCase):
```
Comment (pitrou) [inline numba/tests/test_nrt.py]: If you don't want a test failure to ruin all subsequent tests, you could add `Dummy.alive = 0` in a setUp() method.

File: numba/typing/npydecl.py [hunk 29]
```diff
@@ -318,6 +318,29 @@ def generic(self, args, kws):
...
+class NdEmpty(AbstractTemplate):
+    key = numpy.empty
+
+    def generic(self, args, kws):
+        assert not kws
+        shape = args[0]
+        dtype = types.double
+        if len(args) >= 2:
+            npy_dtype = args[1]
+            dtype = from_dtype(numpy.dtype(npy_dtype.template.key))
```
Comment (pitrou) [inline numba/typing/npydecl.py]: Is there a reason for the complication? Also, I'm wondering: does it work with a structured dtype?

File: numba/runtime/_nrt_python.c (hunk 12 above)
Comment (sklam) [inline numba/runtime/_nrt_python.c]: We do not always defer dtors.

File: numba/runtime/_nrt_python.c (hunk 15 above)
Comment (sklam) [inline numba/runtime/_nrt_python.c]: It is not used in the current code.  Maybe I should get rid of it.

Comment (pitrou) [inline numba/runtime/_nrt_python.c]: Yes, I think getting rid of unused APIs will make the whole thing more palatable :-)

File: numba/runtime/_nrt_python.c (hunk 14 above)
Comment (sklam) [inline numba/runtime/_nrt_python.c]: After some experiment, Py_TPFLAGS_HAVE_NEWBUFFER does not simplify much because numpy and possibly many other packages do not use the check for it.

File: numba/runtime/nrt.c (hunk 18 above)
Comment (sklam) [inline numba/runtime/nrt.c]: atomic_cas also atomically loads. see https://github.com/numba/numba/pull/1077/files#diff-b9938948c081b5f404bb98f9dffa348aR54 and http://llvm.org/docs/LangRef.html#id185

File: numba/runtime/atomicops.py (hunk 22 above)
Comment (sklam) [inline numba/runtime/atomicops.py]: Good catch.

File: numba/tests/test_dyn_array.py (hunk 26 above)
Comment (sklam) [inline numba/tests/test_dyn_array.py]: > Furthermore, they all operate on distinct arrays, so there's not much chance for contention (is there?).
I was testing the CAS for the freelist.  I should add one for the refct as well.

File: numba/typing/npydecl.py (hunk 29 above)
Comment (sklam) [inline numba/typing/npydecl.py]: I will add comments.  It's because things like numpy.int32 is not a dtype but a dtype constructor.

File: numba/tests/test_dyn_array.py (hunk 26 above)
Comment (sklam) [inline numba/tests/test_dyn_array.py]: This first one is actually not very interesting.  I will discard it.

**Comment by sklam** [comment]
Needs to check a reference count problem

File: numba/tests/test_nrt.py (hunk 28 above)
Comment (sklam) [inline numba/tests/test_nrt.py]: Applied at d3ed346

**Comment by sklam** [comment]
Important changes since last review:
* keyword arg `nrt" renamed to "_nrt" and defaults to False
* Buffer object also use NRT if it has a owner object
* Refct initialize to 0 for easier code generation (see https://github.com/sklam/numba/commit/e4e1270aa24635678ba929b7093634454833e44b)
